### PR TITLE
Fixed issue inhabitedtype/angstrom#1

### LINKED ISSUE #1: Use a bigarray directly for the internal buffer
This library uses the [cstruct](https://github.com/mirage/ocaml-cstruct) library for maintaining the parser's internal buffer. I made this decision early on for two reasons:

1. angstrom must maintain two regions in the buffer. The first contains input that has already been consumed by the parser and that can be overridden when provided future input. The second is contains input that has yet to be parsed. A `Cstruct.t` maintains an `off` and `len` field, where the former can be used as the boundary of the two regions. So the necessary bookkeeping was already implemented in the library; and
2. the cstruct library implements copies between `Cstruct.t`s using `memcpy` directly, which is much faster than the default implementation that OCaml provide for bigarrays.

While it was a good idea at the time, cstruct introduces a lot of unnecessary safety checks, it introduces an additional dependency on angstrom, and it introduces an interface that some libraries may be incompatible with.

Therefore, angstrom should drop the dependency on the cstruct library and use bigarrays directly. While doing this, care should be taken to port over the `memcpy` functionality from cstruct. It is very important for performance.

### PR DESCRIPTION
**Pull request by latticetower**
resolves #1

### CODE REVIEW

**Comment by seliopou** [comment]
Hi @latticetower, sorry for the delayed response. I'll take a look at this later today and get back to you with comments.

File: lib/angstrom_stubs.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <sys/param.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <stdint.h>
+
```
Comment (yallop) [inline lib/angstrom_stubs.c]: Are all these `#include`s still necessary?  For example, I don't see anything that uses `<stdint.h>`

Comment (yallop) [inline lib/angstrom_stubs.c]: It looks like only `string.h`, `caml/mlvalues.h` and `caml/bigarray.h` are needed.

Comment (latticetower) [inline lib/angstrom_stubs.c]: @yallop, yes, you are right. I forget to fix them. Thank you for the notice!

**Comment by seliopou** [comment]
Cool! @latticetower, one thing that is more of a nit pick than anything. The Cstruct module that's included should probably be in its own file. I can move that around after merge, or if you'd like to do it, that's great too. Should probably be named something like `Angstrom_cstruct` to avoid name clashes, and be marked an `InternalModule` in the oasis file.

Let me know!

**Comment by latticetower** [comment]
@seliopou, I'll do it

**Comment by dsheets** [comment]
This happened!

**Comment by latticetower** [comment]
@dsheets, yes, it happened. But I also want to add tests and some sort of benchmark for comparision, that's why I asked @seliopou not to merge now.

**Comment by seliopou** [comment]
Closing, see comment in #1.

**Pull request closed without merging**